File: ecubox-backend/src/main/java/com/ecubox/ecubox_backend/entity/EnvioConsolidado.java
package com.ecubox.ecubox_backend.entity;

import com.ecubox.ecubox_backend.enums.EstadoEnvioConsolidadoOperativo;
import com.ecubox.ecubox_backend.enums.EstadoPagoConsolidado;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.EnumSet;

/**
 * Agrupador interno de paquetes para el operario.
 *
 * <p>El estado operativo se persiste en {@link #estadoOperativo}. Los campos
 * {@link #fechaCierre} (cierre de registro), {@link #fechaCerrado} (salida USA) y
 * {@link #fechaArriboEcuador} marcan hitos de cada transición.
 */
@Entity
@Table(name = "envio_consolidado")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class EnvioConsolidado {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "codigo", nullable = false, length = 100)
    private String codigo;

    @Enumerated(EnumType.STRING)
    @Column(name = "estado_operativo", nullable = false, length = 30)
    @Builder.Default
    private EstadoEnvioConsolidadoOperativo estadoOperativo = EstadoEnvioConsolidadoOperativo.VACIO;

    /** Fecha en que se cerró el consolidado para registro (estado CERRADO). */
    @Column(name = "fecha_cierre")
    private LocalDateTime fechaCierre;

    /** Fecha de salida desde USA (estado ENVIADO_DESDE_USA). Antes era el único campo de cierre. */
    @Column(name = "fecha_cerrado")
    private LocalDateTime fechaCerrado;

    /** Fecha de arribo a Ecuador / aduana destino (estado ARRIBADO_ECUADOR). */
    @Column(name = "fecha_arribo_ecuador")
    private LocalDateTime fechaArriboEcuador;

    @Column(name = "peso_total_lbs", precision = 12, scale = 4)
    private BigDecimal pesoTotalLbs;

    @Column(name = "total_paquetes", nullable = false)
    @Builder.Default
    private Integer totalPaquetes = 0;

    /**
     * Estado de pago del envío consolidado. Sincronizado por
     * {@code LiquidacionService} cuando se marca / desmarca pagada la
     * liquidación que contiene a este consolidado en su sección A.
     */
    @Enumerated(EnumType.STRING)
    @Column(name = "estado_pago", nullable = false, length = 20)
    @Builder.Default
    private EstadoPagoConsolidado estadoPago = EstadoPagoConsolidado.NO_PAGADO;

    @Column(name = "created_by")
    private Long createdBy;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @Version
    @Column(nullable = false)
    private Long version;

    private static final java.util.Set<EstadoEnvioConsolidadoOperativo> ESTADOS_CERRADOS =
            EnumSet.of(EstadoEnvioConsolidadoOperativo.CERRADO,
                       EstadoEnvioConsolidadoOperativo.ENVIADO_DESDE_USA,
                       EstadoEnvioConsolidadoOperativo.ARRIBADO_ECUADOR,
                       EstadoEnvioConsolidadoOperativo.RECIBIDO_EN_BODEGA,
                       EstadoEnvioConsolidadoOperativo.LIQUIDADO);

    /** True si el consolidado ya no admite cambios de paquetes (estado posterior a EN_PREPARACION). */
    @Transient
    public boolean isCerrado() {
        return estadoOperativo != null && ESTADOS_CERRADOS.contains(estadoOperativo);
    }

    @Transient
    public boolean isAbierto() {
        return !isCerrado();
    }

    @PrePersist
    void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        if (createdAt == null) createdAt = now;
        if (updatedAt == null) updatedAt = now;
        if (totalPaquetes == null) totalPaquetes = 0;
        if (estadoPago == null) estadoPago = EstadoPagoConsolidado.NO_PAGADO;
        if (estadoOperativo == null) estadoOperativo = EstadoEnvioConsolidadoOperativo.VACIO;
    }

    @PreUpdate
    void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: ecubox-frontend/src/types/envio-consolidado.ts
import type { Paquete } from '@/types/paquete';

/**
 * Envio consolidado: agrupador interno del operario.
 *
 * El estado operativo se representa por `estadoOperativo`; `cerrado` se
 * conserva como bandera compatible de salida USA (`fechaCerrado != null`).
 * No tiene maquina de estados ni se expone en el tracking publico.
 */
export type EstadoPagoConsolidado = 'NO_PAGADO' | 'PAGADO';

/**
 * Estado operativo persistido del consolidado — flujo v2 (V107).
 *
 * Flujo normal: VACIO → EN_PREPARACION → CERRADO → ENVIADO_DESDE_USA
 *   → ARRIBADO_ECUADOR → RECIBIDO_EN_BODEGA → LIQUIDADO
 * Cancelación: VACIO | EN_PREPARACION → CANCELADO
 */
export type EstadoEnvioConsolidadoOperativo =
  | 'VACIO'
  | 'EN_PREPARACION'
  | 'CERRADO'
  | 'ENVIADO_DESDE_USA'
  | 'ARRIBADO_ECUADOR'
  | 'RECIBIDO_EN_BODEGA'
  | 'LIQUIDADO'
  | 'CANCELADO';

/** Un estado dentro del resumen agregado de paquetes de un consolidado. */
export interface EstadoPaqueteResumenItem {
  /** id del estado de rastreo; null representa "Sin estado". */
  estadoId: number | null;
  codigo?: string | null;
  nombre: string;
  cantidad: number;
  ordenTracking?: number | null;
  tipoFlujo?: string | null;
  /** true si requiere atención (flujo alterno o sin estado). */
  requiereAtencion: boolean;
}

/**
 * Resumen agregado de los estados de rastreo de los paquetes de un consolidado.
 * Calculado en backend por lote (sin N+1); solo viene en el listado.
 */
export interface ResumenEstadosPaquetesConsolidado {
  totalPaquetes: number;
  estados: EstadoPaqueteResumenItem[];
  cantidadRequiereAtencion: number;
  estadosMixtos: boolean;
}

export interface EnvioConsolidado {
  id: number;
  codigo: string;
  /** true si el consolidado ya no admite cambios de paquetes (estado posterior a EN_PREPARACION). */
  cerrado: boolean;
  estadoOperativo?: EstadoEnvioConsolidadoOperativo;
  /** Fecha de cierre para registro (estado CERRADO). */
  fechaCierre?: string;
  /** Fecha de salida desde USA (estado ENVIADO_DESDE_USA). */
  fechaCerrado?: string;
  /** Fecha de arribo a Ecuador / aduana destino (estado ARRIBADO_ECUADOR). */
  fechaArriboEcuador?: string;
  pesoTotalLbs?: number;
  totalPaquetes: number;
  /** Estado de pago de la liquidación asociada. */
  estadoPago?: EstadoPagoConsolidado;
  createdAt?: string;
  updatedAt?: string;
  paquetes?: Paquete[];
  /** Resumen agregado de estados de paquetes (solo en el listado paginado). */
  resumenEstadosPaquetes?: ResumenEstadosPaquetesConsolidado;
}

export interface EnvioConsolidadoCreateRequest {
  codigo: string;
  numerosGuia?: string[];
}

export interface EnvioConsolidadoCreateResponse {
  envio: EnvioConsolidado;
  guiasNoEncontradas: string[];
}

export interface EnvioConsolidadoPaquetesRequest {
  paqueteIds: number[];
}

/**
 * @deprecated Usa `PageResponse<EnvioConsolidado>` de `@/types/page` en su lugar.
 *             Se mantiene como alias para no romper consumidores antiguos.
 */
export interface EnvioConsolidadoListResponse {
  content: EnvioConsolidado[];
  totalElements: number;
  totalPages: number;
  page: number;
  size: number;
  first: boolean;
  last: boolean;
}
